### PR TITLE
[runtime] Change merp to emit files for MERP upload mode

### PR DESCRIPTION
**Pull request by alexanderkyte**
An example of the emitted files can be found here
https://gist.github.com/alexanderkyte/47307c95f77ef6e6a2d5efc73cb76dec

It's verbose.

### CODE REVIEW

**Comment by alexanderkyte** [comment]
@monojenkins build pkg

File: mono/mini/mini-exceptions.c [hunk 1]
```diff
@@ -1301,6 +1302,38 @@ mono_get_portable_ip (intptr_t in_ip, intptr_t *out_ip, char *out_name)
...
 }
 
+static intptr_t
+summarize_offset_free_hash (intptr_t accum, MonoFrameSummary *frame)
+{
+	if (!frame->is_managed)
+		return accum;
+
+	// See: mono_ptrarray_hash
+	intptr_t hash_accum = 0;;
```
Comment (lewurm) [inline mono/mini/mini-exceptions.c]: shouldn't this be `intptr_t hash_accum = accum;`?

File: mono/mini/mini-exceptions.c [hunk 2]
```diff
@@ -1301,6 +1302,38 @@ mono_get_portable_ip (intptr_t in_ip, intptr_t *out_ip, char *out_name)
...
+	hash_accum += frame->managed_data.token;
+
+	return hash_accum;
+}
+
+static intptr_t
+summarize_offset_rich_hash (intptr_t accum, MonoFrameSummary *frame)
+{
+	// See: mono_ptrarray_hash
+	intptr_t hash_accum = 0;;
```
Comment (lewurm) [inline mono/mini/mini-exceptions.c]: same here

File: mono/mini/mini-exceptions.c [hunk 3]
```diff
@@ -1301,6 +1302,38 @@ mono_get_portable_ip (intptr_t in_ip, intptr_t *out_ip, char *out_name)
...
+
+	if (frame->is_managed) {
```
Comment (lewurm) [inline mono/mini/mini-exceptions.c]: not sure, but I think the condition needs to be inverted?

File: mono/utils/mono-merp.c [hunk 4]
```diff
@@ -193,151 +195,49 @@ parse_exception_type (const char *signal)
...
+	g_string_append_printf (output, "BlameModuleVersion: %s\n", merp->moduleVersion);
+	g_string_append_printf (output, "BlameModuleOffset: 0x%x\n", merp->moduleOffset);
 
-	print_string_or_blank (output, merp->isObjCException ? "1" : NULL);
+	g_string_append_printf (output, "ExceptionType: %s\n", get_merp_exctype (merp->exceptionArg));
 
-	print_memory_fraction (output, merp->memRes);
-	print_memory_fraction (output, merp->memVirt);
+	g_string_append_printf (output, "StackChecksum: %x\n", merp->hashes.offset_free_hash);
+	g_string_append_printf (output, "StackHash: %x\n", merp->hashes.offset_rich_hash);
```
Comment (lewurm) [inline mono/utils/mono-merp.c]: can we either use `%p` or `0x%08x`/`0x%016x`?

also, what's the reasoning for the naming? we should have it consistent between the output and the names in the struct.

**Comment by lewurm** [comment]
cool idea, I would like to see the stack hash/checksum in non-merp crash output too 🙂

**Comment by alexanderkyte** [comment]
@akoeplinger @luhenry does it seem possible that the failure
```
Unhandled Exception:
System.ExecutionEngineException: System.NullReferenceException: Object reference not set to an instance of an object ---> System.NullReferenceException: Object reference not set to an instance of an object
   --- End of inner exception stack trace ---
  at System.Environment.FailFast (System.String message, System.Exception exception) [0x00007] in <71d8ad678db34313b7f718a414dfcb25>:0 
  at Microsoft.CodeAnalysis.FailFast.OnFatalException (System.Exception exception) [0x00038] in <b9e42f0a228d408b97a82c491aad03ba>:0 
  at Microsoft.CodeAnalysis.FatalError.Report (System.Exception exception, System.Action`1[T] handler) [0x00014] in <b9e42f0a228d408b97a82c491aad03ba>:0 
  at Microsoft.CodeAnalysis.FatalError.Report (System.Exception exception) [0x00000] in <b9e42f0a228d408b97a82c491aad03ba>:0 
  at Microsoft.CodeAnalysis.FatalError.ReportUnlessCanceled (System.Exception exception) [0x0000a] in <b9e42f0a228d408b97a82c491aad03ba>:0 
  at Microsoft.CodeAnalysis.CSharp.MethodCompiler+<>c__DisplayClass21_0.<CompileNamedTypeAsTask>b__0 () [0x0001c] in <c6ef24de369344debe17154b164bb0c5>:0 
  at Roslyn.Utilities.UICultureUtilities+<>c__DisplayClass5_0.<WithCurrentUICulture>b__0 () [0x00038] in <b9e42f0a228d408b97a82c491aad03ba>:0 
  at System.Threading.Tasks.Task.InnerInvoke () [0x0000f] in <71d8ad678db34313b7f718a414dfcb25>:0 
  at System.Threading.Tasks.Task.Execute () [0x00010] in <71d8ad678db34313b7f718a414dfcb25>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <71d8ad678db34313b7f718a414dfcb25>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <71d8ad678db34313b7f718a414dfcb25>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <71d8ad678db34313b7f718a414dfcb25>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.GetResult () [0x00000] in <71d8ad678db34313b7f718a414dfcb25>:0 
  at Microsoft.CodeAnalysis.CSharp.MethodCompiler.WaitForWorkers () [0x00012] in <c6ef24de369344debe17154b164bb0c5>:0 
  at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethodBodies (Microsoft.CodeAnalysis.CSharp.CSharpCompilation compilation, Microsoft.CodeAnalysis.CSharp.Emit.PEModuleBuilder moduleBeingBuiltOpt, System.Boolean emittingPdb, System.Boolean emitTestCoverageData, System.Boolean hasDeclarationErrors, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, System.Predicate`1[T] filterOpt, System.Threading.CancellationToken cancellationToken) [0x00050] in <c6ef24de369344debe17154b164bb0c5>:0 
  at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CompileMethods (Microsoft.CodeAnalysis.Emit.CommonPEModuleBuilder moduleBuilder, System.Boolean emittingPdb, System.Boolean emitMetadataOnly, System.Boolean emitTestCoverageData, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, System.Predicate`1[T] filterOpt, System.Threading.CancellationToken cancellationToken) [0x000aa] in <c6ef24de369344debe17154b164bb0c5>:0 
  at Microsoft.CodeAnalysis.CommonCompiler.CompileAndEmit (Microsoft.CodeAnalysis.TouchedFileLogger touchedFilesLogger, Microsoft.CodeAnalysis.Compilation& compilation, System.Collections.Immutable.ImmutableArray`1[T] analyzers, System.Collections.Immutable.ImmutableArray`1[T] additionalTextFiles, System.Collections.Immutable.ImmutableArray`1[T] embeddedTexts, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, System.Threading.CancellationToken cancellationToken, System.Threading.CancellationTokenSource& analyzerCts, System.Boolean& reportAnalyzer, Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver& analyzerDriver) [0x00216] in <b9e42f0a228d408b97a82c491aad03ba>:0 
  at Microsoft.CodeAnalysis.CommonCompiler.RunCore (System.IO.TextWriter consoleOutput, Microsoft.CodeAnalysis.ErrorLogger errorLogger, System.Threading.CancellationToken cancellationToken) [0x000ef] in <b9e42f0a228d408b97a82c491aad03ba>:0 
  at Microsoft.CodeAnalysis.CommonCompiler.Run (System.IO.TextWriter consoleOutput, System.Threading.CancellationToken cancellationToken) [0x00035] in <b9e42f0a228d408b97a82c491aad03ba>:0 
  at Microsoft.CodeAnalysis.CSharp.CommandLine.Csc+<>c__DisplayClass1_0.<Run>b__0 (System.IO.TextWriter tw) [0x00000] in <038ec9a1d1a64f29a8bc5a3eb89d7e42>:0 
  at Microsoft.CodeAnalysis.CommandLine.ConsoleUtil.RunWithUtf8Output[T] (System.Boolean utf8Output, System.IO.TextWriter textWriter, System.Func`2[T,TResult] func) [0x00034] in <038ec9a1d1a64f29a8bc5a3eb89d7e42>:0 
  at Microsoft.CodeAnalysis.CSharp.CommandLine.Csc.Run (System.String[] args, Microsoft.CodeAnalysis.BuildPaths buildPaths, System.IO.TextWriter textWriter, Microsoft.CodeAnalysis.IAnalyzerAssemblyLoader analyzerLoader) [0x00048] in <038ec9a1d1a64f29a8bc5a3eb89d7e42>:0 
  at (wrapper delegate-invoke) <Module>.invoke_int_string[]_BuildPaths_TextWriter_IAnalyzerAssemblyLoader(string[],Microsoft.CodeAnalysis.BuildPaths,System.IO.TextWriter,Microsoft.CodeAnalysis.IAnalyzerAssemblyLoader)
  at Microsoft.CodeAnalysis.CommandLine.DesktopBuildClient.RunLocalCompilation (System.String[] arguments, Microsoft.CodeAnalysis.BuildPaths buildPaths, System.IO.TextWriter textWriter) [0x00000] in <038ec9a1d1a64f29a8bc5a3eb89d7e42>:0 
  at Microsoft.CodeAnalysis.CommandLine.BuildClient.RunCompilation (System.Collections.Generic.IEnumerable`1[T] originalArguments, Microsoft.CodeAnalysis.BuildPaths buildPaths, System.IO.TextWriter textWriter) [0x00098] in <038ec9a1d1a64f29a8bc5a3eb89d7e42>:0 
  at Microsoft.CodeAnalysis.CommandLine.DesktopBuildClient.Run (System.Collections.Generic.IEnumerable`1[T] arguments, Microsoft.CodeAnalysis.CommandLine.RequestLanguage language, Microsoft.CodeAnalysis.CommandLine.CompileFunc compileFunc, Microsoft.CodeAnalysis.IAnalyzerAssemblyLoader analyzerAssemblyLoader) [0x00043] in <038ec9a1d1a64f29a8bc5a3eb89d7e42>:0 
  at Microsoft.CodeAnalysis.CSharp.CommandLine.Program.MainCore (System.String[] args) [0x00006] in <038ec9a1d1a64f29a8bc5a3eb89d7e42>:0 
  at Microsoft.CodeAnalysis.CSharp.CommandLine.Program.Main (System.String[] args) [0x00000] in <038ec9a1d1a64f29a8bc5a3eb89d7e42>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.ExecutionEngineException: System.NullReferenceException: Object reference not set to an instance of an object ---> System.NullReferenceException: Object reference not set to an instance of an object
   --- End of inner exception stack trace ---
  at System.Environment.FailFast (System.String message, System.Exception exception) [0x00007] in <71d8ad678db34313b7f718a414dfcb25>:0 
  at Microsoft.CodeAnalysis.FailFast.OnFatalException (System.Exception exception) [0x00038] in <b9e42f0a228d408b97a82c491aad03ba>:0 
  at Microsoft.CodeAnalysis.FatalError.Report (System.Exception exception, System.Action`1[T] handler) [0x00014] in <b9e42f0a228d408b97a82c491aad03ba>:0 
  at Microsoft.CodeAnalysis.FatalError.Report (System.Exception exception) [0x00000] in <b9e42f0a228d408b97a82c491aad03ba>:0 
  at Microsoft.CodeAnalysis.FatalError.ReportUnlessCanceled (System.Exception exception) [0x0000a] in <b9e42f0a228d408b97a82c491aad03ba>:0 
  at Microsoft.CodeAnalysis.CSharp.MethodCompiler+<>c__DisplayClass21_0.<CompileNamedTypeAsTask>b__0 () [0x0001c] in <c6ef24de369344debe17154b164bb0c5>:0 
  at Roslyn.Utilities.UICultureUtilities+<>c__DisplayClass5_0.<WithCurrentUICulture>b__0 () [0x00038] in <b9e42f0a228d408b97a82c491aad03ba>:0 
  at System.Threading.Tasks.Task.InnerInvoke () [0x0000f] in <71d8ad678db34313b7f718a414dfcb25>:0 
  at System.Threading.Tasks.Task.Execute () [0x00010] in <71d8ad678db34313b7f718a414dfcb25>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0003e] in <71d8ad678db34313b7f718a414dfcb25>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x00028] in <71d8ad678db34313b7f718a414dfcb25>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x00008] in <71d8ad678db34313b7f718a414dfcb25>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.GetResult () [0x00000] in <71d8ad678db34313b7f718a414dfcb25>:0 
  at Microsoft.CodeAnalysis.CSharp.MethodCompiler.WaitForWorkers () [0x00012] in <c6ef24de369344debe17154b164bb0c5>:0 
  at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethodBodies (Microsoft.CodeAnalysis.CSharp.CSharpCompilation compilation, Microsoft.CodeAnalysis.CSharp.Emit.PEModuleBuilder moduleBeingBuiltOpt, System.Boolean emittingPdb, System.Boolean emitTestCoverageData, System.Boolean hasDeclarationErrors, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, System.Predicate`1[T] filterOpt, System.Threading.CancellationToken cancellationToken) [0x00050] in <c6ef24de369344debe17154b164bb0c5>:0 
  at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CompileMethods (Microsoft.CodeAnalysis.Emit.CommonPEModuleBuilder moduleBuilder, System.Boolean emittingPdb, System.Boolean emitMetadataOnly, System.Boolean emitTestCoverageData, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, System.Predicate`1[T] filterOpt, System.Threading.CancellationToken cancellationToken) [0x000aa] in <c6ef24de369344debe17154b164bb0c5>:0 
  at Microsoft.CodeAnalysis.CommonCompiler.CompileAndEmit (Microsoft.CodeAnalysis.TouchedFileLogger touchedFilesLogger, Microsoft.CodeAnalysis.Compilation& compilation, System.Collections.Immutable.ImmutableArray`1[T] analyzers, System.Collections.Immutable.ImmutableArray`1[T] additionalTextFiles, System.Collections.Immutable.ImmutableArray`1[T] embeddedTexts, Microsoft.CodeAnalysis.DiagnosticBag diagnostics, System.Threading.CancellationToken cancellationToken, System.Threading.CancellationTokenSource& analyzerCts, System.Boolean& reportAnalyzer, Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver& analyzerDriver) [0x00216] in <b9e42f0a228d408b97a82c491aad03ba>:0 
  at Microsoft.CodeAnalysis.CommonCompiler.RunCore (System.IO.TextWriter consoleOutput, Microsoft.CodeAnalysis.ErrorLogger errorLogger, System.Threading.CancellationToken cancellationToken) [0x000ef] in <b9e42f0a228d408b97a82c491aad03ba>:0 
  at Microsoft.CodeAnalysis.CommonCompiler.Run (System.IO.TextWriter consoleOutput, System.Threading.CancellationToken cancellationToken) [0x00035] in <b9e42f0a228d408b97a82c491aad03ba>:0 
  at Microsoft.CodeAnalysis.CSharp.CommandLine.Csc+<>c__DisplayClass1_0.<Run>b__0 (System.IO.TextWriter tw) [0x00000] in <038ec9a1d1a64f29a8bc5a3eb89d7e42>:0 
  at Microsoft.CodeAnalysis.CommandLine.ConsoleUtil.RunWithUtf8Output[T] (System.Boolean utf8Output, System.IO.TextWriter textWriter, System.Func`2[T,TResult] func) [0x00034] in <038ec9a1d1a64f29a8bc5a3eb89d7e42>:0 
  at Microsoft.CodeAnalysis.CSharp.CommandLine.Csc.Run (System.String[] args, Microsoft.CodeAnalysis.BuildPaths buildPaths, System.IO.TextWriter textWriter, Microsoft.CodeAnalysis.IAnalyzerAssemblyLoader analyzerLoader) [0x00048] in <038ec9a1d1a64f29a8bc5a3eb89d7e42>:0 
  at (wrapper delegate-invoke) <Module>.invoke_int_string[]_BuildPaths_TextWriter_IAnalyzerAssemblyLoader(string[],Microsoft.CodeAnalysis.BuildPaths,System.IO.TextWriter,Microsoft.CodeAnalysis.IAnalyzerAssemblyLoader)
  at Microsoft.CodeAnalysis.CommandLine.DesktopBuildClient.RunLocalCompilation (System.String[] arguments, Microsoft.CodeAnalysis.BuildPaths buildPaths, System.IO.TextWriter textWriter) [0x00000] in <038ec9a1d1a64f29a8bc5a3eb89d7e42>:0 
  at Microsoft.CodeAnalysis.CommandLine.BuildClient.RunCompilation (System.Collections.Generic.IEnumerable`1[T] originalArguments, Microsoft.CodeAnalysis.BuildPaths buildPaths, System.IO.TextWriter textWriter) [0x00098] in <038ec9a1d1a64f29a8bc5a3eb89d7e42>:0 
  at Microsoft.CodeAnalysis.CommandLine.DesktopBuildClient.Run (System.Collections.Generic.IEnumerable`1[T] arguments, Microsoft.CodeAnalysis.CommandLine.RequestLanguage language, Microsoft.CodeAnalysis.CommandLine.CompileFunc compileFunc, Microsoft.CodeAnalysis.IAnalyzerAssemblyLoader analyzerAssemblyLoader) [0x00043] in <038ec9a1d1a64f29a8bc5a3eb89d7e42>:0 
  at Microsoft.CodeAnalysis.CSharp.CommandLine.Program.MainCore (System.String[] args) [0x00006] in <038ec9a1d1a64f29a8bc5a3eb89d7e42>:0 
  at Microsoft.CodeAnalysis.CSharp.CommandLine.Program.Main (System.String[] args) [0x00000] in <038ec9a1d1a64f29a8bc5a3eb89d7e42>:0 
make[8]: *** [../../class/lib/basic/mscorlib.dll] Error 1
../../build/library.make:342: recipe for target '../../class/lib/basic/mscorlib.dll' failed```

is actually related to these changes?

**Comment by alexanderkyte** [comment]
I think everything I added is disabled on Linux right now

**Comment by alexanderkyte** [comment]
@monojenkins retry Linux i386

**Comment by luhenry** [comment]
@monojenkins backport 2018-02

**Comment by luhenry** [comment]
@monojenkins backport 2018-04

**Comment by alexanderkyte** [comment]
@luhenry see https://github.com/mono/mono/pull/8870 and https://github.com/mono/mono/pull/8869 . The backport doesn't work because there are other commits that haven't been backported. One of Jay's style fixes need to be backported for the commit to apply cleanly.

**Pull request merged**